ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp

```diff
@@ -0,0 +1,134 @@
+        return tensor.reshape(shape.value);
+    }
+    auto tensor_shape = tensor.shape();
+    auto layout = tensor.layout();
+    auto device = tensor.device();
+    auto memory_config = tensor.memory_config();
+    auto host_tensor = tensor.cpu();
+    auto rm_tensor = ttnn::to_layout(host_tensor, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device *)nullptr);
+    auto host_tensor_4d = unsqueeze_to_4D(rm_tensor);
+    auto tensor_shape_4d = host_tensor_4d.shape();
+    if(tensor_shape.has_tile_padding()) {
+        auto begins = std::vector<uint32_t>({0, 0, 0, 0});
+        auto ends = std::vector<uint32_t>({tensor_shape_4d[0], tensor_shape_4d[1], tensor_shape_4d[2], tensor_shape_4d[3]});
+        auto step = std::vector<uint32_t>({1, 1, 1, 1});
+        host_tensor_4d = ttnn::slice(host_tensor_4d, begins, ends, step, std::nullopt);
+        host_tensor = squeeze_from_4D(host_tensor_4d, tensor_shape.rank());
+    }
+    auto host_reshape_tensor = host_tensor.reshape(shape.value);
+    auto final_layout_tensor = ttnn::to_layout(host_reshape_tensor, layout, std::nullopt, std::nullopt, (Device *)nullptr);
+    auto device_tensor = ttnn::data_transfer_to_device(final_layout_tensor, device, memory_config);
+    return device_tensor;
+}
+
+ttnn::Tensor row_major_reshape(const ttnn::Tensor& tensor, const ttnn::Shape& shape) {
+    const auto layout = tensor.get_layout();
+    auto tensor_shape = tensor.get_shape();
+    if (!ttnn::has_storage_type_of(tensor, ttnn::StorageType::DEVICE)) {
+        return tensor.reshape(shape.value);
+    }
+
+    auto rm_tensor = ttnn::to_layout(tensor, ttnn::ROW_MAJOR_LAYOUT, std::nullopt, std::nullopt, (Device *)nullptr);
+    ttnn::Tensor reshaped_rm_tensor;
+    if (rm_tensor.is_contiguous()) {
+        // Page size depends on the width, so only modify the shape if the width is the same
+        if (tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1]) {
+            return rm_tensor.reshape(shape.value);
+        }
+        //Different page width, need to remake tensor
+        else {
+            uint32_t ROW_MAJOR_WIDTH = 8;
+            auto original_rank = shape.rank();
+
+            if(!(tensor_shape[-1] % ROW_MAJOR_WIDTH == 0 && shape[-1] % ROW_MAJOR_WIDTH == 0)) {
+                return  host_reshape(rm_tensor, shape);
+            }
+            else {
+                auto tensor_4d = unsqueeze_to_4D(rm_tensor);
+                const auto shape_4d = shape.to_rank<4>();
+                auto reshaped_tensor = ttnn::reshape_on_device(tensor_4d, shape_4d[0], shape_4d[1], shape_4d[2], shape_4d[3], tensor.memory_config());
+                reshaped_rm_tensor = squeeze_from_4D(reshaped_tensor, original_rank);
+            }
+        }
+    } else if (tensor_shape.rank() >= 2 and shape.rank() >= 2) {
+        // Handle the case when the tensor is not contiguous but the last two dimensions are the same and so reshape
```

Comment:
is there a test for this case?

Reply:
There is not, I don't think this test is condition is necessary and am unsure how to recreate this (A RM tensor with padding different than legacy shape, I think will make more sense when we pad up to 32 bit alignment with the new tensor) , but I am hesitant to remove it as this is part of the legacy reshape that was in core.cpp before I inherited it.